htlcswitch/link.go

```diff
@@ -384,6 +396,66 @@ type channelLink struct {
 	quit chan struct{}
 }
 
+// hookMap is a data structure that is used to track the hooks that need to be
+// called in various parts of the channelLink's lifecycle.
+//
+// WARNING: NOT thread-safe.
+type hookMap struct {
+	transient  map[uint64]func()
```

Comment:
Given that keys are random and map iteration order is also not guaranteed, could a more deterministic design, such as using a simple slice to store the hooks work better? Or do explicitly want the randomness? If so perhaps worth noting that in the godoc of the type.

Alternatively a priority queue could be considered too if adding hooks is non-deterministic but we still want to ensure order.

Reply:
I'm on the fence. I have arguments for both ways but I'm leaning towards agreeing that a linear design will have fewer surprises for the the caller and I don't have a good enough reason to no promise it. Right now we call the hooks from the control context of the link's main event loop and I think that's a good idea. If we want parallel execution we would just put the logic in its own goroutine and wait on a channel send from the hook. Also while using a slice will cause fragmentation if we continually add and remove persistent hooks there are ways to correct that by tracking the "freed" slots and reallocating them to future adds.

So yeah. Let's promise either queue or stack order on the interface. Do you have a preference?